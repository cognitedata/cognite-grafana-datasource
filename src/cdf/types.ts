export type Datapoints = Items<Datapoint>;
export type Events = Items<Event>;

export interface Datapoint {
  id: number;
  externalId?: string;
  isStep: boolean;
  isString: boolean;
  unit?: string;
  datapoints: TimeSeriesDatapoint[] | TimeSeriesAggregateDatapoint[];
}

export interface TimeSeriesDatapoint {
  timestamp: number;
  value: string;
}

export interface TimeSeriesAggregateDatapoint {
  timestamp: number;
  average?: number;
  max?: number;
  min?: number;
  count?: number;
  sum?: number;
  interpolation?: number;
  stepInterpolation?: number;
  continuousVariance?: number;
  discreteVariance?: number;
  totalVariation?: number;
}

export interface CogniteEvent {
  id: number;
  lastUpdatedTime: string;
  createdTime: string;
  externalId?: string;
  startTime?: number;
  endTime?: number;
  dataSetId?: number;
  description?: string;
  type?: string;
  subtype?: string;
  assetIds?: number[];
  source?: string;
  sourceId?: string;
  metadata?: { [s: string]: string };
}

export interface FilterRequestParams {
  metadata?: Metadata;
  assetSubtreeIds?: IdEither[];
  createdTime?: TimeRange;
  lastUpdatedTime?: TimeRange;
  externalIdPrefix?: string;
}

export interface AssetsFilterRequestParams extends FilterRequestParams {
  name?: string;
  parentIds?: CogniteInternalId[];
  parentExternalIds?: CogniteExternallId[];
  rootIds?: IdEither[];
  source?: string;
  root?: boolean;
}

export interface TimeseriesFilterRequestParams extends FilterRequestParams {
  name?: string;
  unit?: string;
  isString?: boolean;
  isStep?: boolean;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
}

export interface EventsFilterRequestParams extends FilterRequestParams {
  startTime?: TimeRange;
  endTime?: TimeRange;
  activeAtTime?: TimeRange;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
  source?: string;
  type?: string;
  subtype?: string;
}

export type EventsFilterTimeParams =
  | Pick<EventsFilterRequestParams, 'activeAtTime'>
  | Pick<EventsFilterRequestParams, 'startTime' | 'endTime'>;

export interface FilterRequest<Filter> extends Limit, Cursor {
  filter?: Filter;
  advancedFilter?: any;
}

export interface Resource {
  id: number;
  externalId?: string;
  name?: string;
  description?: string;
}

export interface TimeSeriesResponseItem extends Resource {
  isString?: boolean;
  metadata?: Record<string, string>;
  unit?: string;
  assetId?: string;
  isStep: boolean;
  source?: string;
  sourceId?: string;
  createdTime: number;
  lastUpdatedTime: number;
  selected: boolean;
}

export type TimeSeriesResponse = Items<TimeSeriesResponseItem>;

export type Limit = {
  limit?: number;
};

export interface Cursor {
  cursor?: string;
}

export interface Metadata {
  [name: string]: string;
}

export type TimeRange = Range<number>;
export type CogniteInternalId = number;
export type CogniteExternallId = string;

export interface Timestamp {
  timestamp: number;
}

export type Items<T = any> = {
  items: T[];
};

export type IdEither =
  | {
      id: number;
    }
  | {
      externalId: string;
    };

export interface Range<T> {
  min?: T;
  max?: T;
}

export interface CogniteLabelsResponse {
  externalId: string;
  name: string;
  description: string;
  dataSetId: number;
  createdTime: number;
}
interface CogniteRelationshipAsset {
  createdTime?: number;
  lastUpdatedTime?: number;
  rootId?: number;
  aggregates?: {
    childCount: number;
    depth: number;
    path?: [
      {
        id: string | number;
      }
    ];
  };
  parentId?: number;
  parentExternalId?: string;
  externalId: string;
  name?: string;
  description?: string;
  dataSetId: number;
  metadata?: { [s: string]: string };
  source?: string;
  labels?: CogniteLabelsResponse[];
  id: number;
}
export interface CogniteRelationshipResponse {
  externalId: string;
  sourceExternalId: string;
  sourceType?: string;
  targetExternalId: string;
  targetType?: string;
  startTime?: number;
  endTime?: number;
  confidence?: number;
  dataSetId?: number;
  labels?: CogniteLabelsResponse[];
  createdTime?: number;
  lastUpdatedTime?: number;
  source?: CogniteRelationshipAsset;
  target?: CogniteRelationshipAsset;
}
export interface RelationshipsFilter {
  dataSetIds?: {
    id: number;
    value?: string;
  }[];
  labels?: {
    containsAny: {
      externalId: string;
      value?: string;
    }[];
  };
  isActiveAtTime?: boolean;
  activeAtTime?: {
    max: number;
    min: number;
  };
  sourceExternalIds?: string[];
  targetTypes?: string[];
}
