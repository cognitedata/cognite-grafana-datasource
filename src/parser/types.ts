import { STSClientFilter } from './ts';

export const FilterTypes = {
  RegexNotEquals: '!~' as FilterType,
  RegexEquals: '=~' as FilterType,
  NotEquals: '!=' as FilterType,
  Equals: '=' as FilterType,
};
export type FilterType = '!~' | '=~' | '!=' | '=';

export enum QueryParserTypes {
  assets = 'assets',
  events = 'events',
}

export interface QueryParserType {
  type: QueryParserTypes;
}

export interface QueryParserResponse extends QueryParserType {
  query: QueryCondition[];
}

export type ParsedFilter = STSClientFilter;

export interface QueryCondition {
  [key: string]: any;
}

export interface ParserResponse extends QueryParserType {
  params: QueryCondition;
  filters: ParsedFilter[];
}
