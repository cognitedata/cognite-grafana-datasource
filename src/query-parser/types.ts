export const FilterType = {
  RegexNotEquals: '!~' as FilterType,
  RegexEquals: '=~' as FilterType,
  NotEquals: '!=' as FilterType,
  Equals: '=' as FilterType,
};
export type FilterType = '!~' | '=~' | '!=' | '=';
export interface QueryFilter {
  filter: FilterType;
  value: string;
}

export enum QueryParserTypes {
  assets = 'assets',
  events = 'events',
}

export interface QueryParserType {
  type: QueryParserTypes;
}

export interface QueryParserResponse extends QueryParserType {
  query: QueryCondition[];
}

export interface ParsedFilter extends QueryFilter {
  path: string;
}

export interface QueryCondition {
  [key: string]: any;
}

export interface ParserResponse extends QueryParserType {
  params: QueryCondition;
  filters: ParsedFilter[];
}
