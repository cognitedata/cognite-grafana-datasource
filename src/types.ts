import {
  DataQueryOptions,
  DataQuery,
  TimeSeries,
  TimeRange as GrafanaTimeRange,
  DataSourceSettings,
} from '@grafana/ui';

export function isError(maybeError: DataQueryError | any): maybeError is DataQueryError {
  return (<DataQueryError>maybeError).error !== undefined;
}

/**
 * Comes from grafana, could be imported in future releases hopefully
 * @param name â€“ event name
 */
export const eventFactory = <T = undefined>(name: string): AppEvent<T> => {
  return { name };
};

export type QueryResponse = DataResponse<TimeSeries[]>;

export interface MetricDescription {
  readonly text: string;
  readonly value: number | string;
}

export type MetricFindQueryResponse = MetricDescription[];

export enum Tab {
  Timeseries = 'Timeseries',
  Asset = 'Asset',
  Custom = 'Custom',
}

export enum ParseType {
  Timeseries = 'Timeseries',
  Asset = 'Asset',
  Event = 'Event',
}

export interface TimeSeriesResponseItem {
  id: number;
  externalId?: string;
  name?: string;
  isString?: boolean;
  metadata?: object;
  unit?: string;
  assetId?: string;
  isStep: boolean;
  description?: string;
  source?: string;
  sourceId?: string;
  createdTime: number;
  lastUpdatedTime: number;
  selected: boolean;
}

export type TimeSeriesResponse = Items<TimeSeriesResponseItem>;

export interface AssetQuery {
  target: string;
  includeSubtrees: boolean;
  old?: AssetQuery;
  timeseries?: TimeSeriesResponseItem[];
  func?: string;
  templatedTarget?: string;
}

export interface InputQueryTarget extends DataQuery {
  target: number | '';
  aggregation: string;
  granularity: string;
  error: string;
  label: string;
  tab: Tab;
  assetQuery: AssetQuery;
  expr: string;
  warning: string;
}

export interface QueryTarget extends InputQueryTarget {
  target: number;
}

export type QueryFormat = 'json';

export type QueryOptions = DataQueryOptions<InputQueryTarget>;

export type Tuple<T> = [T, T];

export interface Range<T> {
  min?: T;
  max?: T;
}

/**
 * Comes from grafana, could be imported in future releases hopefully
 */
export interface AppEvent<T> {
  readonly name: string;
  payload?: T;
}

export enum HttpMethod {
  POST = 'POST',
  GET = 'GET',
  PATCH = 'PATCH',
  DELETE = 'DELETE',
}

export interface DataSourceRequestOptions {
  url: string;
  method: HttpMethod;
  retry?: number;
  requestId?: string;
  headers?: { [s: string]: string };
  silent?: boolean;
  data?: object;
}

export interface Timestamp {
  timestamp: number;
}

export type MetaResponses = Responses<ResponseMetadata, DataQueryRequestResponse>;

export type SuccessResponse<Metadata, Response> = { metadata: Metadata; result: Response };
export type FailResponse<Metadata> = { metadata: Metadata; error: any };
export type Responses<Metadata, Response> = {
  failed: FailResponse<Metadata>[];
  succeded: SuccessResponse<Metadata, Response>[];
};

export interface TimeSeriesDatapoint extends Timestamp {
  value: string;
}

export interface TimeSeriesAggregateDatapoint extends Timestamp {
  average?: number;
  max?: number;
  min?: number;
  count?: number;
  sum?: number;
  interpolation?: number;
  stepInterpolation?: number;
  continuousVariance?: number;
  discreteVariance?: number;
  totalVariation?: number;
}

export interface Datapoint {
  id: number;
  externalId?: string;
  isStep: boolean;
  isString: boolean;
  unit?: string;
  datapoints: TimeSeriesDatapoint[] | TimeSeriesAggregateDatapoint[];
}

export type Datapoints = Items<Datapoint>;

export interface DataQueryRequestResponse extends DataResponse<Datapoints> {
  config: {
    data: {
      aggregates: string;
      limit: number;
    };
  };
}

export interface RequestParams<DataType = any> {
  path: string;
  data: DataType;
  method: HttpMethod;
  params?: { [s: string]: any };
  requestId?: string;
  playground?: boolean;
  cacheTime?: string;
}

export type DataQueryError = {
  error: {
    cancelled?: boolean;
    data: {
      error?: {
        message: string;
        notFound?: string[];
      };
    };
    status: number;
  };
};

export type QueriesData = {
  items: DataQueryRequestItem[];
  target: QueryTarget;
}[];

export type ResponseMetadata = { labels: string[]; target: QueryTarget };

export interface DataQueryAlias {
  alias: string;
  id: number;
  aggregate?: string;
  granularity?: string;
}

export type IdEither =
  | {
      id: number;
    }
  | {
      externalId: string;
    };

export type DataQueryRequestItem = {
  expression?: string;
  start?: string | number;
  end?: string | number;
  limit?: number;
  granularity?: string;
  aggregates?: string[];
  id?: number;
};

export type Aggregates = Pick<DataQueryRequest, 'aggregates'>;
export type Granularity = Pick<DataQueryRequest, 'granularity'>;

export interface DataQueryRequest {
  items: DataQueryRequestItem[];
  start?: string | number;
  end?: string | number;
  limit?: number;
  aggregates?: string[];
  granularity?: string;
}

export interface Annotation {
  datasource: string;
  enable: boolean;
  hide: boolean;
  iconColor: string;
  limit: number;
  name: string;
  query: string;
  filter: string;
  error: string;
  type: string;
  tags: string[];
}

export interface AnnotationQueryOptions {
  range: GrafanaTimeRange;
  rangeRaw: GrafanaTimeRange;
  annotation: Annotation;
  dashboard: number;
}

export interface AnnotationResponse {
  annotation: Annotation;
  title: string;
  time: number;
  timeEnd?: number;
  text: string;
  tags?: string[];
  isRegion?: boolean;
}

export interface AnnotationSearchQuery {
  description: string;
  type: string;
  subtype: string;
  minStartTime: number;
  maxStartTime: number;
  minEndTime: number;
  maxEndTime: number;
  minCreatedTime: number;
  maxCreatedTime: number;
  minLastUpdatedTime: number;
  maxLastUpdatedTime: number;
  // format is {"k1": "v1", "k2": "v2"}
  metadata: string;
  assetIds: number[];
  assetSubtrees: number[];
  sort: 'startTime' | 'endTime' | 'createdTime' | 'lastUpdatedTime';
  dir: 'asc' | 'desc';
  limit: number;
  offset: 0;
}

export interface Event {
  id: number;
  startTime: number;
  endTime: number;
  description: string;
  type: string;
  subtype: string;
  assetIds: number[];
  source: string;
  sourceId: string;
}

export type Events = Items<Event>;

export type DataEvents = DataResponse<Events>;

export type AnnotationQueryRequestResponse = DataResponse<DataEvents>;

export interface DataResponse<T> {
  data: T;
}

export type Items<T = object> = {
  items: T[];
};

export type CursorResponse<T> = DataResponse<Items<T> & { nextCursor?: string }>;

export type Response<T = object> = DataResponse<{
  items: T[];
}>;

export type TimeseriesFilterQuery = {
  filter?: {
    description?: string;
    assetSubtreeIds?: IdEither[];
    assetIds?: string[];
    [s: string]: any; // so we auto-support next features
  };
  cursor?: string;
} & Limit;

export type Limit = {
  limit?: number;
};

export interface Cursor {
  cursor?: string;
}

export interface VariableQueryData {
  query: string;
  error?: string;
}

export interface VariableQueryProps {
  query: string;
  onChange: (query: VariableQueryData) => void;
  datasource: any;
  templateSrv: any;
}

export interface CogniteDataSourceSettings extends DataSourceSettings {
  jsonData: {
    authType: string;
    defaultRegion: string;
    cogniteProject: string;
  };
}

export interface Metadata {
  [name: string]: string;
}

export type TimeRange = Range<number>;
export type CogniteInternalId = number;
export type CogniteExternallId = string;

export interface FilterRequestParams {
  metadata?: Metadata;
  assetSubtreeIds?: IdEither[];
  createdTime?: TimeRange;
  lastUpdatedTime?: TimeRange;
  externalIdPrefix?: string;
}

export interface AssetsFilterRequestParams extends FilterRequestParams {
  name?: string;
  parentIds?: CogniteInternalId[];
  parentExternalIds?: CogniteExternallId[];
  rootIds?: IdEither[];
  source?: string;
  root?: boolean;
}

export interface TimeseriesFilterRequestParams extends FilterRequestParams {
  name?: string;
  unit?: string;
  isString?: boolean;
  isStep?: boolean;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
}

export interface EventsFilterRequestParams extends FilterRequestParams {
  startTime?: TimeRange;
  endTime?: TimeRange;
  assetIds?: CogniteInternalId[];
  assetExternalIds?: CogniteExternallId[];
  rootAssetIds?: IdEither[];
  source?: string;
  type?: string;
  subtype?: string;
}

export interface FilterRequest<Filter> extends Limit, Cursor {
  filter: Filter;
}

export interface QueryRequestError {
  refId: string;
  error: string;
}

export interface QueryDatapointsLimitWarning {
  refId: string;
  warning: string;
}
