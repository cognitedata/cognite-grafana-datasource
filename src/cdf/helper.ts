import _ from 'lodash';

export const filterLabels = (labels) =>
  !_.isEmpty(_.get(labels, 'containsAny')) && {
    labels: {
      containsAny: _.map(labels.containsAny, ({ value }) => ({ externalId: value })),
    },
  };

export const filterExternalId = (sourceExternalIds, isTypeTimeseries: boolean) =>
  !_.isEmpty(sourceExternalIds) && {
    targetTypes: isTypeTimeseries ? ['timeSeries'] : undefined,
    sourceExternalIds,
  };
export const filterdataSetIds = (dataSetIds) =>
  !_.isEmpty(dataSetIds) && {
    dataSetIds: _.map(dataSetIds, ({ value }) => ({ id: Number(value) })),
  };
