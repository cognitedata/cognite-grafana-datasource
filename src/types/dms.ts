// DMS API types
export interface DMSSpace {
  space: string;
  name?: string;
  description?: string;
  createdTime: number;
  lastUpdatedTime: number;
}

export interface DMSView {
  space: string;
  externalId: string;
  version: string;
  name?: string;
  description?: string;
  createdTime: number;
  lastUpdatedTime: number;
}

export interface DMSInstanceId {
  space: string;
  externalId: string;
}

export interface DMSFilter {
  and?: DMSFilter[];
  or?: DMSFilter[];
  not?: DMSFilter;
  equals?: {
    property: string[];
    value: any;
  };
  in?: {
    property: string[];
    values: any[];
  };
  range?: {
    property: string[];
    gte?: any;
    gt?: any;
    lte?: any;
    lt?: any;
  };
}

export interface DMSSearchRequest {
  view: {
    type: 'view';
    space: string;
    externalId: string;
    version: string;
  };
  query?: string;
  filter?: DMSFilter;
  limit?: number;
  properties?: string[];
}

export interface DMSSearchResponse {
  items: DMSInstance[];
}

export interface DMSInstance {
  instanceType: 'node' | 'edge';
  space: string;
  externalId: string;
  version: number;
  lastUpdatedTime: number;
  createdTime: number;
  deletedTime?: number;
  properties?: {
    [space: string]: {
      [view: string]: {
        [property: string]: any;
      };
    };
  };
}
