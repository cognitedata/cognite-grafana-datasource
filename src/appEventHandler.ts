import { AppEvent, EventBusSrv } from '@grafana/data';
import { getCalculationWarnings, getLimitsWarnings, stringifyError } from './cdf/client';
import { failedResponseEvent, responseWarningEvent } from './constants';
import { SuccessResponse } from './types';

export const eventBusService = new EventBusSrv();

export async function emitEvent<T>(event: AppEvent<T>, payload: T): Promise<void> {
  return eventBusService.emit(event, payload);
}

export function handleError(error: any, refId: string) {
  const errMessage = stringifyError(error);
  emitEvent(failedResponseEvent, { refId, error: errMessage });
}
export function handleWarning(warningMessage: string, refId: string) {
  emitEvent(responseWarningEvent, { refId, warning: warningMessage });
}
export function showWarnings(responses: SuccessResponse[]) {
  responses.forEach(({ result, metadata }) => {
    const { items } = result.data;
    const { limit } = result.config.data;
    const { refId } = metadata.target;
    const warning = [getLimitsWarnings(items, limit), getCalculationWarnings(items)]
      .filter(Boolean)
      .join('\n\n');

    if (warning) {
      emitEvent(responseWarningEvent, { refId, warning });
    }
  });
}
