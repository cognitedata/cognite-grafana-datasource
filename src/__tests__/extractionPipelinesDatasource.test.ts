/* eslint-disable no-nested-ternary */
import _ from 'lodash';
import { CogniteQuery, Tab } from '../types';
import { getMockedDataSource } from '../test_utils';
import { ExtractionPipelinesResponse } from '../cdf/types';
import { eventBusService } from '../appEventHandler';

type Mock = jest.Mock;
type QueryTargetLike = Partial<CogniteQuery>;

jest.mock('@grafana/data');

const appEvents = eventBusService;

const { ExtractionPipelines } = Tab;
const ds = getMockedDataSource();
const { backendSrv } = ds;
const options: any = {
  targets: [],
  range: {
    from: 1549336675000,
    to: 1549338475000,
  },
  interval: '30s',
  format: 'json',
  panelId: 1,
  dashboardId: 1,
};
const extpipeError = {
  status: 400,
  data: {
    error: {
      code: 400,
      message: 'extpipe error message',
    },
  },
};
const extpipeErrorMessage = {
  refId: 'A',
  error: '[400 ERROR] extpipe error message',
};
const noSelectedRunsError = {
  refId: 'A',
  error: '[ERROR] Please select value for extraxtion pipelines runs',
};
const columns = ['name', 'externalId', 'runId', 'status'];
const baseTarget: QueryTargetLike = {
  tab: ExtractionPipelines,
  refId: 'A',
};
const extpipesRes = {
  items: [
    { name: 'Test 1', externalId: 'test-1' },
    { name: 'Test 2', externalId: 'test-2' },
    { name: 'Test 3', externalId: 'test-3' },
  ],
};
const extpipesByIdRes = {
  items: [{ name: 'Test 1', externalId: 'test-1' }],
};
const extpipesRunsListRes1 = {
  items: [
    { id: 1, status: 'seen' },
    { id: 2, status: 'failure' },
    { id: 3, status: 'succsess' },
  ],
};
const extpipesRunsListRes2 = {
  items: [
    { id: 10, status: 'succsess' },
    { id: 20, status: 'seen' },
    { id: 30, status: 'failure' },
  ],
};
const extpipesRunsListRes3 = {
  items: [
    { id: 100, status: 'failure' },
    { id: 200, status: 'succsess' },
    { id: 300, status: 'seen' },
  ],
};
const extpipeRunsValue = ({ items }) => _.map(items[0], (value) => value);
const extpipeRunsValues = ({ items }) => _.map(items, (item) => _.map(item, (value) => value));
const extpipesWithRunsValues = ({ items }) =>
  _.map(items, (item: ExtractionPipelinesResponse, index: number) =>
    _.concat(
      [item.name, item.externalId],
      extpipeRunsValue(
        index === 1
          ? extpipesRunsListRes2
          : index === 2
          ? extpipesRunsListRes3
          : extpipesRunsListRes1
      )
    )
  );
describe('extraction pipelines', () => {
  let results;
  beforeEach(() => {
    jest.clearAllMocks();
  });
  describe('getExtractionPipelinesDropdowns', () => {
    beforeEach(async () => {
      backendSrv.datasourceRequest = jest
        .fn()
        .mockImplementationOnce(() => Promise.resolve({ data: extpipesRes }))
        .mockResolvedValueOnce((x) => Promise.resolve(x.data));
      results = await ds.extractionPipelinesDatasource.getExtractionPipelinesDropdowns('a');
    });
    it('it is called 1 times', () => {
      expect(backendSrv.datasourceRequest).toBeCalledTimes(1);
    });
    it('returns response data', () => {
      expect(results).toEqual(extpipesRes.items);
    });
  });
  describe('query', () => {
    let runsList;
    describe('on empty selected', () => {
      describe('with runs', () => {
        beforeEach(async () => {
          options.targets = [
            {
              ...baseTarget,
              extractionPipelinesQuery: {
                columns,
                selections: [],
                getRuns: true,
              },
            },
          ];
          results = await ds.query(options);
        });
        it('emits error', () => {
          expect(appEvents.emit).toHaveBeenCalledTimes(1);
        });
        it('error message', () => {
          expect((appEvents.emit as Mock).mock.calls[0][1]).toEqual(noSelectedRunsError);
        });
        it('empty tabledata', () => {
          expect(results.data[0].rows).toEqual([]);
        });
        it('column', () => {
          expect(results.data[0].columns).toEqual(columns.map((text) => ({ text })));
        });
      });
      describe('no runs', () => {
        beforeAll(async () => {
          options.targets = [
            {
              ...baseTarget,
              extractionPipelinesQuery: {
                columns,
                selections: [],
                getRuns: false,
              },
            },
          ];
          backendSrv.datasourceRequest = jest
            .fn()
            .mockImplementationOnce(() => Promise.resolve({ data: extpipesRes }))
            .mockImplementationOnce(() => Promise.resolve({ data: extpipesRunsListRes1 }))
            .mockImplementationOnce(() => Promise.resolve({ data: extpipesRunsListRes2 }))
            .mockImplementationOnce(() => Promise.resolve({ data: extpipesRunsListRes3 }))
            .mockImplementation((x) => Promise.resolve(x.data));
          results = await ds.query(options);
          runsList = extpipesWithRunsValues(extpipesRes);
        });
        it('dont emits error', () => {
          expect(appEvents.emit).toHaveBeenCalledTimes(0);
        });
        it('columns', () => {
          expect(results.data[0].columns).toEqual(columns.map((text) => ({ text })));
        });
        it('table data', () => {
          expect(results.data[0].rows).toEqual(runsList);
        });
        it('match snapshot', () => {
          expect(results.data).toMatchSnapshot();
        });
      });
    });
    describe('on selected', () => {
      const extpipeRunsColumns = ['id', 'status'];
      describe('no runs', () => {
        beforeEach(async () => {
          options.targets = [
            {
              ...baseTarget,
              extractionPipelinesQuery: {
                columns,
                selections: [{ value: 'test-1', label: 'Test 1' }],
                getRuns: false,
              },
            },
          ];
          backendSrv.datasourceRequest = jest
            .fn()
            .mockImplementationOnce(() => Promise.resolve({ data: extpipesByIdRes }))
            .mockImplementationOnce(() => Promise.resolve({ data: extpipesRunsListRes1 }))
            .mockImplementation((x) => Promise.resolve(x.data));
          results = await ds.query(options);
          runsList = extpipesWithRunsValues(extpipesByIdRes);
        });
        it('dont emits error', () => {
          expect(appEvents.emit).toHaveBeenCalledTimes(0);
        });
        it('table data', () => {
          expect(results.data[0].rows).toEqual(runsList);
        });
        it('columns', () => {
          expect(results.data[0].columns).toEqual(columns.map((text) => ({ text })));
        });
        it('match snapshot', () => {
          expect(results.data).toMatchSnapshot();
        });
      });
      describe('with runs', () => {
        const extpipesRunsListRes = [
          extpipesRunsListRes1,
          extpipesRunsListRes2,
          extpipesRunsListRes3,
        ];
        beforeEach(async () => {
          options.targets = [
            {
              ...baseTarget,
              extractionPipelinesQuery: {
                columns: extpipeRunsColumns,
                selections: [
                  { value: 'test-1', label: 'Test 1' },
                  { value: 'test-2', label: 'Test 2' },
                  { value: 'test-3', label: 'Test 3' },
                ],
                getRuns: true,
              },
            },
          ];
          backendSrv.datasourceRequest = jest
            .fn()
            .mockImplementationOnce(() => Promise.resolve({ data: extpipesRunsListRes1 }))
            .mockImplementationOnce(() => Promise.resolve({ data: extpipesRunsListRes2 }))
            .mockImplementationOnce(() => Promise.resolve({ data: extpipesRunsListRes3 }))
            .mockImplementation((x) => Promise.resolve(x.data));
          results = await ds.query(options);
        });
        it('dont emits error', () => {
          expect(appEvents.emit).toHaveBeenCalledTimes(0);
        });
        it(`has results data length of ${extpipesRunsListRes.length}`, () => {
          expect(results.data.length).toEqual(3);
        });
        it('has column, table-name', () => {
          for (let i = 0; i < results.data.length; i += 1) {
            runsList = extpipeRunsValues(extpipesRunsListRes[i]);
            expect(results.data[i].columns).toEqual(extpipeRunsColumns.map((text) => ({ text })));
            expect(results.data[i].rows).toEqual(runsList);
            expect(results.data[i].name).toEqual(`Test ${i + 1}`);
          }
        });
        it('match snapshot', () => {
          expect(results.data).toMatchSnapshot();
        });
      });
    });
  });
});
