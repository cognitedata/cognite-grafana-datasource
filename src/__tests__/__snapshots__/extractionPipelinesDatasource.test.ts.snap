// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extraction pipelines query on empty selected no runs match snapshot 1`] = `
Array [
  Object {
    "columns": Array [
      Object {
        "text": "name",
      },
      Object {
        "text": "externalId",
      },
      Object {
        "text": "runId",
      },
      Object {
        "text": "status",
      },
    ],
    "name": "Extraction Pipelines",
    "rows": Array [
      Array [
        "Test 1",
        "test-1",
        1,
        "seen",
      ],
      Array [
        "Test 2",
        "test-2",
        10,
        "succsess",
      ],
      Array [
        "Test 3",
        "test-3",
        100,
        "failure",
      ],
    ],
    "type": "table",
  },
]
`;

exports[`extraction pipelines query on selected no runs match snapshot 1`] = `
Array [
  Object {
    "columns": Array [
      Object {
        "text": "name",
      },
      Object {
        "text": "externalId",
      },
      Object {
        "text": "runId",
      },
      Object {
        "text": "status",
      },
    ],
    "name": "Extraction Pipelines",
    "rows": Array [
      Array [
        "Test 1",
        "test-1",
        1,
        "seen",
      ],
    ],
    "type": "table",
  },
]
`;

exports[`extraction pipelines query on selected with runs match snapshot 1`] = `
Array [
  Object {
    "columns": Array [
      Object {
        "text": "id",
      },
      Object {
        "text": "status",
      },
    ],
    "name": "test-1",
    "rows": Array [
      Array [
        1,
        "seen",
      ],
      Array [
        2,
        "failure",
      ],
      Array [
        3,
        "succsess",
      ],
    ],
    "type": "table",
  },
  Object {
    "columns": Array [
      Object {
        "text": "id",
      },
      Object {
        "text": "status",
      },
    ],
    "name": "test-2",
    "rows": Array [
      Array [
        10,
        "succsess",
      ],
      Array [
        20,
        "seen",
      ],
      Array [
        30,
        "failure",
      ],
    ],
    "type": "table",
  },
  Object {
    "columns": Array [
      Object {
        "text": "id",
      },
      Object {
        "text": "status",
      },
    ],
    "name": "test-3",
    "rows": Array [
      Array [
        100,
        "failure",
      ],
      Array [
        200,
        "succsess",
      ],
      Array [
        300,
        "seen",
      ],
    ],
    "type": "table",
  },
]
`;
